cloudinit/config/schema.py

```diff
@@ -109,6 +157,106 @@ def is_schema_byte_string(checker, instance):
     ) or isinstance(instance, (bytes,))
 
 
+def _add_deprecation_msg(description: Optional[str] = None):
+    msg = "DEPRECATED."
+    if description:
+        msg += f" {description}"
+    return msg
+
+
+def _validator_deprecated(
+    _validator,
+    deprecated: bool,
+    _instance,
+    schema: dict,
+    error_type: Type[Exception],
+):
+    """Jsonschema validator for `deprecated` items.
+
+    It raises a instance of `error_type` if deprecated that must be handled,
+    otherwise the instance is consider faulty.
+    """
+    if deprecated:
+        description = schema.get("description")
+        msg = _add_deprecation_msg(description)
+        yield error_type(msg)
+
+
+def _anyOf(validator, anyOf, instance, _schema, error_type: Type[Exception]):
```

Comment:
Rather than writing a custom validator implementations, could we extend the existing functions instead?

Poking around at how other projects extend jsonschema validators, there appears to be a general pattern that allows using the existing jsonschema validators and adding extra functionality:

```
def _extend_oneOf(self, validator_class):
        validate_oneOf = validator_class.VALIDATORS["oneOf"]

        def custom_oneOf(validator, properties, instance, schema):
            for error in validate_oneOf(
                validator, properties, instance, schema,
            ):
                yield error

            # add custom functionality here
        return validators.extend(
            validator_class, {"oneOf": custom_oneOf},
        ) 
```

Since the `.extend()` interface is intended for this use case, I'd like to see us use it so we don't have to worry about drift from upstream's validator function implementations, additional test coverage for basic jsonschema functionality, etc.

Reply:
That would be nice, but we cannot extend them in that way as:

Let's suppose we have the following schema and instance of it

```py
schema = {
    "$schema": "http://json-schema.org/draft-04/schema#",
    "properties": {
        "delay": {
            "oneOf": [
                {"type": "integer", "description": "Minutes of delay."},
                {
                    "type": "string",
                    "deprecated": True,
                    "description": "DEPRECATED. Use an integer.",
                },
            ]
        },
    },
}

instance = {"delay": "+5"}
```

Then if extend `onOf` as suggested, `validate_oneOf` wouldn't yield any error as the instance is valid and `custom_oneOf` wouldn't be able to determine to which sub-schema `delay` belongs to. And we should not emit false-positive deprecation warning.

Solving the above problem implies to have a tight integration with `oneOf`, `anyOf`, etc. As they do not offer an API / hooks to do so, the only way to do so (I think) is to override them.

Note: A similar test case covering this requirement is added here: https://github.com/canonical/cloud-init/pull/1549/files#diff-ccf5aac2a6067d8a0782d74b172805f80e71225a212822e4f4e158f9ca1e6e3aR514

Reply:
Sorry for the delayed response on this. +1 Thanks for the explanation!